### PR TITLE
Reduce memory usage and makes saving constraints optional

### PR DESCRIPTION
**Pull request by bqpd**
- uses sets of varkeys instead of KeySets in many places
    - deletes KeySets after use for string-substitution during ConstraintSet creation
 - deletes substitution attributes after used for mmaping / dual calculation

Reduces memory consumption for large models with string substitutions during creation by ~30%

For future reference: consider not using KeySets for such one-offs at all, instead doing a one-pass over varkeys. Honestly such iteration is cheap enough that KeySets might be removed entirely, and KeyDicts used only for solutions. This wouldn't lead to large changes in speed/memory, but it would remove a lot of code.

Also makes saving constraints with a solution optional

### CODE REVIEW

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test models please

**Comment by bqpd** [comment]
test this please

**Comment by bqpd** [comment]
test models please

File: gpkit/constraints/set.py [hunk 1]
```diff
@@ -64,58 +64,60 @@ class ConstraintSet(list, ReprMixin):
     "Recursive container for ConstraintSets and Inequalities"
     unique_varkeys, idxlookup = frozenset(), {}
     _name_collision_varkeys = None
+    _varkeys = None
 
-    def __init__(self, constraints, substitutions=None):  # pylint: disable=too-many-branches,too-many-statements
+    def __init__(self, constraints, substitutions=None, *, bonusvks=None):  # pylint: disable=too-many-branches,too-many-statements
```
Comment (whoburg) [inline gpkit/constraints/set.py:69]: docstring for `bonusvks`? Or at the very least an inline comment?

File: gpkit/constraints/gp.py [hunk 2]
```diff
@@ -305,33 +305,33 @@ def _compile_result(self, solver_out):
...
-        self.v_ss = cost_senss.copy()
+        gpv_ss = cost_senss.copy()
         for las, nus, c in zip(la[1:], self.nu_by_posy[1:], self.hmaps[1:]):
             while getattr(c, "parent", None) is not None:
                 c = c.parent
             v_ss, c_senss = c.sens_from_dual(las, nus, result)
             for vk, x in v_ss.items():
-                self.v_ss[vk] = x + self.v_ss.get(vk, 0)
-            while getattr(c, "generated_by", None) is not None:
+                gpv_ss[vk] = x + gpv_ss.get(vk, 0)
```
Comment (whoburg) [inline gpkit/constraints/gp.py:314]: haven't thought deeply about this, but seems like this is acting a lot like a `defaultdict`? Could `gpv_ss` be a default dict, and then this simply becomes `gpv_ss[vk] += x`? Is `defaultdict` noticeably slower?

File: gpkit/constraints/gp.py [hunk 3]
```diff
@@ -305,33 +305,33 @@ def _compile_result(self, solver_out):
...
+            while getattr(c, "generated_by", None):
                 c = c.generated_by
```
Comment (whoburg) [inline gpkit/constraints/gp.py:316]: inline comment here? Something like "loop to top of hierarchy"?

File: gpkit/constraints/gp.py [hunk 4]
```diff
@@ -305,33 +305,33 @@ def _compile_result(self, solver_out):
...
+            dlogcost_dlogv = gpv_ss.pop(v)
             val = np.array(result["constants"][v])
             for c, dv_dc in v.gradients.items():
                 with warnings.catch_warnings():  # skip pesky divide-by-zeros
                     warnings.simplefilter("ignore")
                     dlogv_dlogc = dv_dc * result["constants"][c]/val
-                    before = self.v_ss.get(c, 0)
-                    self.v_ss[c] = before + dlogcost_dlogv*dlogv_dlogc
+                    before = gpv_ss.get(c, 0)
+                    gpv_ss[c] = before + dlogcost_dlogv*dlogv_dlogc
```
Comment (whoburg) [inline gpkit/constraints/gp.py:327]: same comment as above, seems like `defaultdict` might simplify this?

File: gpkit/keydict.py [hunk 5]
```diff
@@ -45,7 +45,7 @@ class KeyMap:
     collapse_arrays = False
     keymap = []
     log_gets = False
-    varkeys = None
+    cset = None
```
Comment (whoburg) [inline gpkit/keydict.py:48]: documentation / inline comment for `cset`? 😬

File: gpkit/constraints/set.py [hunk 6]
```diff
@@ -64,58 +64,60 @@ class ConstraintSet(list, ReprMixin):
...
-                    del key.descr["value"]
-                    if key.veckey and key.veckey.value is not None:
-                        del key.veckey.descr["value"]
+            self._varkeys = None
+        for key in self.vks:
+            if key not in self.substitutions:
+                if key.veckey is None or key.veckey not in self.substitutions:
+                    continue
+                if np.isnan(self.substitutions[key.veckey][key.idx]):
+                    continue
```
Comment (whoburg) [inline gpkit/constraints/set.py:104]: these `continue` cases seem complex enough that an inline comment would be beneficial

File: gpkit/constraints/set.py [hunk 7]
```diff
@@ -64,58 +64,60 @@ class ConstraintSet(list, ReprMixin):
...
+            self.bounded.add((key, "upper"))
+            self.bounded.add((key, "lower"))
+            if key.value is not None and not key.constant:
+                del key.descr["value"]
+                if key.veckey and key.veckey.value is not None:
+                    del key.veckey.descr["value"]
```
Comment (whoburg) [inline gpkit/constraints/set.py:110]: ?? I've read this a few times and still confused what's going on here... inline comment? (sorry broken record)

File: gpkit/nomials/data.py [hunk 8]
```diff
@@ -18,9 +18,6 @@ class NomialData(ReprMixin):
 
     def __init__(self, hmap):
         self.hmap = hmap
-        self.vks = set()
-        for exp in self.hmap:
-            self.vks.update(exp)
```
Comment (whoburg) [inline gpkit/nomials/data.py:23]: 🎉

File: gpkit/nomials/map.py [hunk 9]
```diff
@@ -152,18 +152,16 @@ def mmap(self, orig):
...
         selfexps = list(self.keys())
         for orig_exp, self_exp in self.expmap.items():
             if self_exp not in self:  # can occur in tautological constraints
                 continue              #     after substitution
             fraction = self.csmap.get(orig_exp, orig[orig_exp])/self[self_exp]
-            m_from_ms[self_exp][orig_exp] = fraction
             orig_idx = origexps.index(orig_exp)
             pmap[selfexps.index(self_exp)][orig_idx] = fraction
-        return pmap, m_from_ms
+        return pmap
```
Comment (whoburg) [inline gpkit/nomials/map.py:164]: 🎉

**Review by whoburg — APPROVED** [review]

**Comment by bqpd** [comment]
test models please

**Pull request merged**